### ISSUE TITLE
Float and Int zeros can be added to Values

### ISSUE THREAD

**Issue by zchen088** [labrad/pylabrad#140]
In [39]: 0.0+1*ns
Out[39]: Value(1.0, 'ns')

In [40]: 0.1+1*ns
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-40-081e8407e5cb> in <module>()
```
----> 1 0.1+1*ns

C:\Users\Jimmy\Desktop\pylabrad\labrad\units.pyc in __add__(self, other)
    257
    258     def __add__(self, other):
--> 259         return WithUnit(self._sum(other, 1, 1), self.unit)
    260
    261     __radd__ = __add__

C:\Users\Jimmy\Desktop\pylabrad\labrad\units.pyc in _sum(self, other, sign1, si
n2)
    254             return sign1 * self._value
    255         else:
--> 256             raise TypeError("Incompatible Units %s, ''" % (self.unit,))
    257
    258     def __add__(self, other):

TypeError: Incompatible Units ns, ''
> c:\users\jimmy\desktop\pylabrad\labrad\units.py(256)_sum()
    255         else:
--> 256             raise TypeError("Incompatible Units %s, ''" % (self.unit,))
    257
```

**Comment by zchen088** [comment]
Here's a python quirk I had never seen before:
```
In [56]: x = 1

In [57]: id(x)
Out[57]: 30515928

In [58]: x += 1

In [59]: id(x)
Out[59]: 30515916

In [60]: x += 0

In [61]: id(x)
Out[61]: 30515916

In [62]: x = x+0

In [63]: id(x)
Out[63]: 30515916
```

**Comment by DanielSank** [comment]
@ejeffrey I remember you talking about wanting to add this feature. Thinking about it now I'm not sure it's useful, and it seems like it could make certain operations succeed by accident where they should fail.

**Comment by ejeffrey** [comment]
Yes, this feature is deliberate, if possibly ill advised.  The idea was
that I wanted to be able to compare to use expressions like:

if dt > 0:
   do stuff

without having to specify units on dt, since 0 ns == 0 s == 0 ks.  This
expression has a clear, unambiguous interpretation, and adding anything to
it tends to make it harder to understand, so I thought it was worth
supporting.  If it causes other problems, maybe it should be removed.  It
is also nice when calling functions that expect a Value to be able to pass
zero if the number is irrelevant or unused.  It kind of makes sense because
units work somewhat like multiplication, so you might expect that 0*ns == 0
should be true.  As you can see, this special case only works for zero, and
only works for "normal" units, not things like C / F that have offsets.

With regards to the id behavior: this is only true for -128 <= x <= 127 or
so.  This is an internal cpython optimization that has singletons for the
most common small integers.  So integers in that range will always have the
same id().  That won't work in general for arbitrary integers.

**Comment by DanielSank** [comment]
> With regards to the id behavior: this is only true for -128 <= x <= 127 or so.  This is an internal cpython optimization that has singletons for the most common small integers.  So integers in that range will always have the same id().  That won't work in general for arbitrary integers.

TIL

I'm kind of for removing the ability to add e.g. `0` and `0*ns`. I don't think
```python
if dt > 0*ns:
    do stuff
```
is any harder to understand than without the unit, and the following makes me nervous
```python
x = 2*ns
y = 2*ns
x - y == 0
>>> True
```

**Comment by ejeffrey** [comment]
Maybe I am insufficiently creative, but I can't imagine what someone would
type "x - y == 0" and expect it to do anything other than "x == y".  I am
OK with surprising people by accidentally doing what they want, just not
doing the wrong thing when we could raise an error.  If someone can come up
with a realistic case where someone would actually make a bug that
enforcing units would solve I am fine with changing it.  Be careful, it is
kind of baked in right now because I think the comparison operators
internally turn 'x<y' into 'x-y < 0'

Here is an (admittedly contrived) example of where you might want to
compare to zero but not care what the units are:

def quadratic(a,b,c):
   discriminant = b**2 - 4*a*c
   if discriminant > 0:
       return (-b + sqrt(discriminant))/(2*a),  (-b -
sqrt(discriminant))/(2*a)
   elif discriminant == 0:
       return -b/(2*a)
   else:
       raise ValueError

>>> quadratic(1*m/s**2, 4*m/s, 3*m)
(Value(-1.0, 's'), Value(-3.0, 's'))

Obviously in this case you could just convert the comparison to b**2 >
4*a*c, but the point is that quadratic doesn't need to know anything about
units to work properly.

**Comment by DanielSank** [comment]
The contrived example compares floats for equality. Seems dodgy. That said, so does my example, so probably it's all a wash.

**Comment by ejeffrey** [comment]
OK, well you can have it return double roots for the equality case if you
don't like it.

**Comment by DanielSank** [comment]
wut?

**Comment by ejeffrey** [comment]
Not relevant to the matter at hand, but the only reason I did the
comparison for equality was so that I could return a single value when the
roots were degenerate.  If you just change 'if discriminant > 0' to 'if
discriminant >= 0' you can drop the equality clause and be fine.  But that
is beside the point which was not to write the best quadratic equation
solver, but to illustrate that it might be helpful to compare something to
zero when you don't know what units it has.

**Comment by DanielSank** [comment]
Ok so what's the plan here?

**Comment by ejeffrey** [comment]
Did this cause a bug?  If this is just Jimmy wondering if something is
intentional, then I vote close as WAI.  If it caused a bug in pyle or
other, then we should try to remove it and see how much (if any) code in
pyle breaks.

**Comment by DanielSank** [comment]
He raised this because there was a suspicious line in `gates.py` with `0` passed as the start time as an envelope. I think this is a works-as-intended (WAI) but I'll leave it to @zchen088 to close if he thinks it's appropriate.

**Comment by maffoo** [comment]
I would be in favor of changing this behavior. I think having a special case for comparison to 0 goes against the spirit of dimensional safety that the units library was meant to provide in the first place. You can still do generic comparisons to 0 if you want: `x[x.unit] > 0` or `x > 0*x.unit`. It's a little trickier to write a function that does comparison to 0 and works for both values with units and bare floats, but I think such a function is probably a bad idea anyway. Being too flexible in what types we accept seems to be a source of lots of the problems in our code.

Even if we decide to continue allowing comparison to zero, I think we should rework the implementation to not allow addition of bare floats; that seems like an unintended consequence of the way comparisons was implemented.
